hibernate-core/src/main/java/org/hibernate/sql/Template.java

```diff
@@ -205,6 +206,40 @@ else if ( quotedIdentifier && ( dialect.closeQuote()==token.charAt(0) ) ) {
 				continue;
 			}
 
+			// Special processing for type casting
+			if ( inCast && "as".equals( lcToken ) ) {
+				result.append( " as " );
+				//to manage cases like cast(field as numeric(10,4))
+				int openBrackets = 0;
+
+				hasMore = tokens.hasMoreTokens();
+				nextToken = hasMore ? tokens.nextToken() : null;
+
+				while ( hasMore && inCast ) {
+					if ( ")".equals( nextToken ) ) {
```

Comment:
I was a little confused about why  openBrackets == 0 before decrementing, but now I realize that openBrackets was not incremented on the first '('. 

I tried using this formula, "cast( 0.88 * cast( cents as number(10,4) ) / 100 as number(10,0) )", and it did not cause an error, but the inCast was set to false at the end of the inner cast (not at the end of the outer cast).

It seems like other functions will be broken as suggested in HHH-9951. If there's a more general way to deal with functions it would be worth re-working this.